### PR TITLE
e2e_part2

### PR DESCRIPTION
**Pull request by AYAHASSAN287**
## Summary

End-to-end test coverage for the wrapper send path, plus the supporting infrastructure (helpers, CI pipeline, store assertions). Adds 19 new scenarios across two files

### Tests 

`tests/wrappers_tests/test_send_e2e_part1.py` (`TestSendBeforeRelay`):

| ID | Scenario |
|----|----------|
| S17 | Send before any relay peer joins; positive/negative paths |
| S19 | Store peer comes online after propagation — Sent fires once reachable |
| S20 | Relay/lightpush propagation succeeds, store query initially misses, retry republishes and validates |
| S21 | Delivery retry window expires before any valid path recovers |
| S22 | Non-ephemeral message with `reliabilityEnabled=False` — Propagated only, no Sent |
| S23 | Non-ephemeral, reliability enabled, no store peer reachable — Propagated only, no Sent, no terminal error |
| S24 | Ephemeral message with reachable store peer — Propagated only, no store validation |
| S26 | Lightpush peer churn — selected peer disappears, alternate remains, message still propagates |
| S30 | Concurrent sends on the same content topic during initial auto-subscribe — unique request IDs, no cross-association |
| S31 | Concurrent sends across mixed topics during 3-phase peer churn (before / during / after restart) |

`tests/wrappers_tests/test_send_e2e_part2.py`:

| ID | Scenario |
|----|----------|
| S02 | Send without explicit subscribe — auto-subscribe on first send |
| S06 | Core sender, relay only — propagation without store |
| S07 | Core sender, relay + store — propagation with store validation |
| S08 | Relay-to-lightpush fallback |
| S09 | Relay-to-lightpush fallback with store validation |
| S10 | Edge sender, lightpush only |
| S12 | Isolated sender, no peers — error event |
| S14 | Lightpush non-retryable error (oversized message) |
| S15 | Lightpush retryable error then recovery |

### New helpers
- `src/node/wrapper_helpers.py` — `EventCollector` (thread-safe async event sink), event-type predicates, `wait_for_propagated/sent/error/connected`, `assert_event_invariants`, `get_node_multiaddr`, `create_message_bindings`.
- `src/steps/store.py` — `check_sent_message_is_stored` to assert hashes are present in store.
- `src/node/wrappers_manager.py` — `get_node_info_raw` (returns the unparsed string).

### Infra / wiring
- New `.github/workflows/pr_tests.yml` — build (cached on submodule hash) → wrapper tests → smoke; full 18-shard suite triggered by `full-test` label.

## Known issues
- **S19** __ failing in the post-propagation reliability/store phase: the message is propagated, but the sender cannot complete store validation before the retry window expires although Store peer joins and is detected:``
`storeClientCount=1 `  [issue](https://github.com/logos-messaging/logos-delivery/issues/3848)
- **S31** — `json_serialization/writer.nim → w.inObject() AssertionDefect` and nwaku SIGSEGV during shutdown/churn. Note: flaky not show up always 
- **S20** — when the store peer joins after the initial miss, the retry loop fails to republish: sender repeatedly logs `NoPeersToPublish` and eventually emits `message_error: Unable to send within retry time window`.
   note: the main point is that it's not possible to force the miss round 
- **S08**__ the test fails if lightpushnode not used  [issue](https://github.com/logos-messaging/logos-delivery/issues/3847)

### CODE REVIEW

**Review by fbarbu15 — COMMENTED** [review]
LGTM, thanks but can you please add a full run test report to the PR description?

File: .github/workflows/pr_tests.yml [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+          path: vendor/logos-delivery-python-bindings/lib/
+
+      - name: Run wrapper tests - basic life cycle
+        env:
+          PYTHONPATH: ${{ github.workspace }}/vendor/logos-delivery-python-bindings/waku
+        run: |
+          pytest tests/wrappers_tests/test_basic_life_cycle.py \
+            -m "not docker_required" \
+            --reruns 2 \
+            --junit-xml=wrapper-results-basic.xml || true
```
Comment (at0m1x19) [inline .github/workflows/pr_tests.yml]: The trailing `|| true` here (and at lines 168, 177) swallows pytest's exit code, so the job stays green even when tests actually fail. Could we remove it and rely on proper `xfail`/`skip` markers for tests we know are red?

File: tests/wrappers_tests/test_send_e2e_part2.py [hunk 2]
```diff
@@ -0,0 +1,703 @@
...
+                "staticnodes": [service_addr],
+                "portsShift": 1,
+            }
+            relay_result = WrapperManager.create_and_start(config=relay_config)
+            assert relay_result.is_ok(), f"Failed to start relay peer: {relay_result.err()}"
+
+            with relay_result.ok_value:
+                sender_config = {
+                    **node_config,
+                    # "lightpushnode": service_addr, #this comment currently raise issue
```
Comment (at0m1x19) [inline tests/wrappers_tests/test_send_e2e_part2.py:302]: With `lightpushnode` commented out, no `staticnodes`, and `discv5Discovery=False`, the sender doesn't seem to have a path to publish — yet the test still asserts `Propagated`. Per #163 S08 the setup expects "at least one lightpush peer reachable for the target shard". Would it make sense to either uncomment `lightpushnode` (and dig into the underlying issue) or skip the test until that's resolved?

File: tests/wrappers_tests/test_send_e2e_part2.py [hunk 3]
```diff
@@ -0,0 +1,703 @@
...
+                "lightpush": False,
+                "store": False,
+                "staticnodes": [service_addr],
+                "portsShift": 1,
+            }
+            relay_result = WrapperManager.create_and_start(config=relay_config)
+            assert relay_result.is_ok(), f"Failed to start relay peer: {relay_result.err()}"
+
+            with relay_result.ok_value:
+                sender_config = {
```
Comment (at0m1x19) [inline tests/wrappers_tests/test_send_e2e_part2.py]: Heads up: `sender_config = {**node_config, ...}` inherits `store: True` from the service's `node_config` above, so the sender ends up with both a local store and a remote `storenode` — which weakens the S09 premise (per #163, the store-validation should be coming from the **remote** store peer). Would adding `"store": False` to `sender_config` make sense?

File: src/node/wrapper_helpers.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+    if EVENT_SENT in first_index and EVENT_PROPAGATED in first_index:
+        assert first_index[EVENT_PROPAGATED] < first_index[EVENT_SENT], (
+            f"message_sent (index {first_index[EVENT_SENT]}) arrived before "
+            f"message_propagated (index {first_index[EVENT_PROPAGATED]}) "
+            f"for request {request_id}. Events: {events}"
+        )
+
+
+def get_node_multiaddr(node) -> str:
```
Comment (at0m1x19) [inline src/node/wrapper_helpers.py:176]: `get_node_multiaddr` `.strip()`s and returns whatever `MyMultiaddresses` returns. If the wrapper ever returns more than one address (newline/comma/JSON list), this would silently break. Could we at least assert a single line?

**Review by at0m1x19 — COMMENTED** [review]
A few comments — mostly suggestions, nothing blocking the overall direction. Thanks for the work on this!

File: tests/wrappers_tests/test_send_e2e_part1.py [hunk 5]
```diff
@@ -0,0 +1,1002 @@
...
+                    collector=sender_collector,
+                    request_id=request_id,
+                    timeout_s=NO_SENT_OBSERVATION_S,
+                )
+                assert sent_event is not None, (
+                    f"No MessageSentEvent received within {NO_SENT_OBSERVATION_S}s "
+                    f"from a store-enabled relay peer. Collected events: {sender_collector.events}"
+                )
+
+    # @pytest.mark.xfail(reason="scenario might be not possible to simulate")
```
Comment (at0m1x19) [inline tests/wrappers_tests/test_send_e2e_part1.py]: A consolidated question — what's the plan with the commented-out `xfail` markers here (and at lines 225, 841 for S20 / S31)? The PR description acknowledges these as known-failing, and the marker is sitting right there commented out. Either we believe they should pass (and the comments can be deleted), or they're known-failing (and the marker should be active with `strict=False` + a tracking issue link). The current state — broken tests plus commented-out markers — gives us the worst of both worlds: CI fails, and there's no readable signal in the source about why.

File: tests/wrappers_tests/test_send_e2e_part1.py [hunk 6]
```diff
@@ -0,0 +1,1002 @@
...
+    def test_s19_store_peer_appears_after_propagation(self, node_config):
```
Comment (at0m1x19) [inline tests/wrappers_tests/test_send_e2e_part1.py:137]: Validated against the latest completed PR run (25327035558): `Wrapper Tests` runs with `-m "not docker_required"`, but these tests (S19 here, also S20 at line 226 and S31 at line 842) still get picked up because the marker isn't applied. S19 fails on the spec'd assertion, S20 fails earlier with `400 Bad Request ... admin/v1/peers`. Could we tag these with `@pytest.mark.docker_required` so the wrapper-only job filters them out cleanly?

File: tests/wrappers_tests/test_send_e2e_part1.py [hunk 7]
```diff
@@ -0,0 +1,1002 @@
+from concurrent.futures import ThreadPoolExecutor
+from time import time_ns
+
+import pytest
+from src.env_vars import NODE_2
+from src.steps.common import StepsCommon
```
Comment (at0m1x19) [inline tests/wrappers_tests/test_send_e2e_part1.py:5]: A few imports here look unused — could we drop them?
- L2 `from time import time_ns`
- L4 `import pytest` (only referenced in the commented-out `@pytest.mark.xfail` lines)
- L6 `from src.steps.common import StepsCommon` (the test class extends `StepsStore`, not `StepsCommon` directly)

File: tests/wrappers_tests/test_send_e2e_part2.py [hunk 8]
```diff
@@ -0,0 +1,703 @@
...
+    EventCollector,
+    assert_event_invariants,
+    create_message_bindings,
+    get_node_multiaddr,
+    wait_for_connected,
+    wait_for_propagated,
+    wait_for_sent,
+    wait_for_error,
+)
+from src.steps.store import StepsStore
```
Comment (at0m1x19) [inline tests/wrappers_tests/test_send_e2e_part2.py]: Two unused imports here:
- L3 `import pytest`
- L18 `from src.steps.store import StepsStore` (none of the test classes extend it)

**Review by at0m1x19 — COMMENTED** [review]
Follow-up batch — a question on the xfail markers, a validated note on docker_required, and a couple of unused imports.

**Review by at0m1x19 — COMMENTED** [review]
Final batch of part1 comments

File: tests/wrappers_tests/test_send_e2e_part1.py [hunk 9]
```diff
@@ -0,0 +1,1002 @@
...
+                        collector=sender_collector,
+                        request_id=request_id,
+                        timeout_s=0,
+                    )
+                    assert error_event is None, f"Unexpected message_error for request_id={request_id}: {error_event}"
+
+                # Cross-association guard: every event with a requestId must
+                # belong to exactly one of the request ids we issued.
+                issued = set(request_ids)
+                for event in sender_collector.events:
```
Comment (at0m1x19) [inline tests/wrappers_tests/test_send_e2e_part1.py]: Small concern: `sender_collector.events` is iterated here without `_lock` while `event_callback` is appending from another thread, which can raise `RuntimeError: list changed size during iteration` or skip events. Would it make sense to snapshot the list under the lock first (e.g. via `get_events_for_request` or a similar helper)?

File: tests/wrappers_tests/test_send_e2e_part1.py [hunk 10]
```diff
@@ -0,0 +1,1002 @@
...
+            for request_id in phase2_ids:
+                error_event = wait_for_error(
+                    collector=sender_collector,
+                    request_id=request_id,
+                    timeout_s=0,
+                )
+                assert error_event is None, f"Unexpected terminal message_error for phase-2 " f"request_id={request_id} after recovery: {error_event}"
+
+            issued = set(all_request_ids)
+            for event in sender_collector.events:
```
Comment (at0m1x19) [inline tests/wrappers_tests/test_send_e2e_part1.py]: Same unlocked iteration as in S30 above — could we apply the same fix here?

File: tests/wrappers_tests/test_send_e2e_part1.py [hunk 11]
```diff
@@ -0,0 +1,1002 @@
...
+                self.check_published_message_is_stored(
+                    store_node=store_node,
+                    pubsub_topic=self.test_pubsub_topic,
+                    messages_to_check=[message],
+                    page_size=5,
+                    ascending="true",
+                )
+
+    # @pytest.mark.xfail(reason="error:NoPeersToPublish")
+    def test_s20_store_misses_initially_then_retry_succeeds(self, node_config):
```
Comment (at0m1x19) [inline tests/wrappers_tests/test_send_e2e_part1.py:233]: The setup here doesn't quite match #163 S20 — there's no store peer at T0, so an "initial store query miss" can't really happen; this is closer to S19 ("store peer absent initially, store peer appears later"). #163 S20 reads as "first propagation succeeds, but when store validation starts the message is not yet in store; later retries republish the message and a store peer eventually archives it" — i.e. store reachable from T0 but archive empty. To exercise S20, could we keep a store-capable peer reachable from T0 but defer it actually archiving (e.g. delay its relay subscription until after the first Propagated)?

File: tests/wrappers_tests/test_send_e2e_part1.py [hunk 12]
```diff
@@ -0,0 +1,1002 @@
...
+        relay_peer.start(relay="true", discv5_discovery="false")
+        relay_peer.set_relay_subscriptions([self.test_pubsub_topic])
+
+        lightpush_peer = WakuNode(NODE_2, f"s31_lightpush_peer_{self.test_id}")
+        lightpush_peer.start(relay="true", lightpush="true", discv5_discovery="false")
+        lightpush_peer.set_relay_subscriptions([self.test_pubsub_topic])
+
+        store_peer = WakuNode(NODE_2, f"s31_store_peer_{self.test_id}")
+        store_peer.start(relay="true", store="true", discv5_discovery="false")
+        store_peer.set_relay_subscriptions([self.test_pubsub_topic])
```
Comment (at0m1x19) [inline tests/wrappers_tests/test_send_e2e_part1.py:862]: Three docker `WakuNode`s are started here without `try/finally` or a fixture, so any later assertion failure can leak containers across runs. Same pattern in S19 (`store_node` line 197) and S20 (`store_node` line 321). Could we wrap them in a fixture or `try/finally`?

File: tests/wrappers_tests/test_send_e2e_part1.py [hunk 13]
```diff
@@ -0,0 +1,1002 @@
...
+                    continue
+                assert event_request_id in issued, (
+                    f"Event carries an unknown requestId={event_request_id!r}, " f"not in issued set {issued}. Event: {event}"
+                )
+
+            # Use the hash the wrapper emitted on message_sent so the store
+            # lookup matches the exact bytes that were actually published.
+            phase3_hashes = []
+            for request_id in phase3_ids:
+                sent_event = wait_for_sent(
```
Comment (at0m1x19) [inline tests/wrappers_tests/test_send_e2e_part1.py:970]: This wait is on store validation rather than propagation, so I think `SENT_AFTER_STORE_TIMEOUT_S=60s` would be a better fit than `PROPAGATED_TIMEOUT_S=30s` — 30s after docker churn for 8 messages might be flaky.

File: tests/wrappers_tests/test_send_e2e_part1.py [hunk 14]
```diff
@@ -0,0 +1,1002 @@
...
+                    timeout_s=PROPAGATED_TIMEOUT_S,
+                )
+                assert propagated_event is not None, (
+                    f"No MessagePropagatedEvent received within {PROPAGATED_TIMEOUT_S}s "
+                    f"after relay peer joined. Collected events: {sender_collector.events}"
+                )
+                sent_event = wait_for_sent(
+                    collector=sender_collector,
+                    request_id=request_id,
+                    timeout_s=NO_SENT_OBSERVATION_S,
```
Comment (at0m1x19) [inline tests/wrappers_tests/test_send_e2e_part1.py]: Small constant mismatch: `wait_for_sent(timeout_s=NO_SENT_OBSERVATION_S=5s)` is asserting Sent **does** fire, but `NO_SENT_OBSERVATION_S` is the "expect no Sent" window. Could we use `SENT_TIMEOUT_S` (10s) here, like S07 does? 5s also feels a bit tight for store validation right after the relay peer joins.

File: tests/wrappers_tests/test_send_e2e_part1.py [hunk 15]
```diff
@@ -0,0 +1,1002 @@
...
+    "/test/1/s31-topic-d/proto",
+    "/test/1/s31-topic-e/proto",
+    "/test/1/s31-topic-f/proto",
+    "/test/1/s31-topic-g/proto",
+    "/test/1/s31-topic-h/proto",
+]
+
+
+class TestSendBeforeRelay(StepsStore):
+    def test_s17_send_before_relay_peers_joins(self, node_config):
```
Comment (at0m1x19) [inline tests/wrappers_tests/test_send_e2e_part1.py:63]: `assert_event_invariants` is applied to every test in part2 but to none in part1. Per #163's "Assertions Worth Capturing In Every Simulation Test", the same invariants apply universally (same `RequestId`, no duplicate terminal events, `Sent` after `Propagated`) — including retry-path scenarios. Could we add the helper to part1 tests too? In S20 / S31 in particular, if the implementation does emit duplicate Propagated on retry, that'd be exactly the kind of regression the helper is meant to catch.

File: tests/wrappers_tests/test_send_e2e_part1.py [hunk 16]
```diff
@@ -0,0 +1,1002 @@
...
+                    f"Unexpected MessageSentEvent for an ephemeral message. "
+                    f"Ephemeral messages must never be store-validated.\n"
+                    f"Sent event: {sent_event}\n"
+                    f"Collected events: {sender_collector.events}"
+                )
+
+    def test_s26_lightpush_peer_churn_alternate_remains(self, node_config):
+        """
+        S26: multiple lightpush peers, the selected one disappears,
+             an alternate remains.
```
Comment (at0m1x19) [inline tests/wrappers_tests/test_send_e2e_part1.py:630]: The docstring says "Two lightpush server peers", but the topology is actually three peers (peer1 + relay_peer + peer2). Could we expand the comment to clarify the role of `relay_peer` (kept alive as the fallback gossipsub mesh after peer1 disappears?)?

**Comment by AYAHASSAN287** [comment]
> LGTM, thanks but can you please add a full run test report to the PR description?

[latest job](https://github.com/logos-messaging/logos-delivery-interop-tests/actions/runs/25365842699)
here is already 2 open bugs related to the open issues

File: .github/workflows/pr_tests.yml (hunk 1 above)
Comment (AYAHASSAN287) [reply .github/workflows/pr_tests.yml]: done

File: tests/wrappers_tests/test_send_e2e_part2.py (hunk 2 above)
Comment (AYAHASSAN287) [reply tests/wrappers_tests/test_send_e2e_part2.py:302]: for this point I discussed it with Ivan . we may mark it as failing test for now instead of bypassing it .
I will add xfail mark

File: tests/wrappers_tests/test_send_e2e_part2.py (hunk 3 above)
Comment (AYAHASSAN287) [reply tests/wrappers_tests/test_send_e2e_part2.py]: done

File: src/node/wrapper_helpers.py (hunk 4 above)
Comment (AYAHASSAN287) [reply src/node/wrapper_helpers.py:176]: yes , assertion added

File: tests/wrappers_tests/test_send_e2e_part1.py (hunk 6 above)
Comment (AYAHASSAN287) [reply tests/wrappers_tests/test_send_e2e_part1.py:137]: .the tests here need to be  part of the wrapper suite maybe we shoud not strict the wrapper only job to exclude all docket tests

File: tests/wrappers_tests/test_send_e2e_part2.py (hunk 8 above)
Comment (AYAHASSAN287) [reply tests/wrappers_tests/test_send_e2e_part2.py]: L3 now used
L18 removed

File: tests/wrappers_tests/test_send_e2e_part1.py (hunk 7 above)
Comment (AYAHASSAN287) [reply tests/wrappers_tests/test_send_e2e_part1.py:5]: L4 used now 
L6 removed

File: tests/wrappers_tests/test_send_e2e_part1.py (hunk 9 above)
Comment (AYAHASSAN287) [reply tests/wrappers_tests/test_send_e2e_part1.py]: addressed

File: tests/wrappers_tests/test_send_e2e_part1.py (hunk 10 above)
Comment (AYAHASSAN287) [reply tests/wrappers_tests/test_send_e2e_part1.py]: done

File: tests/wrappers_tests/test_send_e2e_part1.py (hunk 11 above)
Comment (AYAHASSAN287) [reply tests/wrappers_tests/test_send_e2e_part1.py:233]: this scenario is realy hard to simulate , It was already mentioned by Zoltan that we may have some scenarios that  can't be simulated  , I changed it again but still not match so I skipped it with comment as not possible to simulate
 When the store node started with --relay=true --shard=0, it auto-subscribed to /waku/2/rs/198/0 during boot, before any test code ran so no way to delay the connection

File: tests/wrappers_tests/test_send_e2e_part1.py (hunk 12 above)
Comment (AYAHASSAN287) [reply tests/wrappers_tests/test_send_e2e_part1.py:862]: the root conftest tears those down after each test, so containers shouldn't leak across runs even when an assertion fails mid-test. That's why none of the existing tests in this file (S19, S20, S26, S29, etc.) wrap node starts in try/finally ,it would be redundant with the existing teardown.

File: tests/wrappers_tests/test_send_e2e_part1.py (hunk 13 above)
Comment (AYAHASSAN287) [reply tests/wrappers_tests/test_send_e2e_part1.py:970]: I'd rather not reuse SENT_AFTER_STORE_TIMEOUT_S here — it specifically means "waiting for sent confirmation after a store peer joined late," I used another one with 45s

File: tests/wrappers_tests/test_send_e2e_part1.py (hunk 14 above)
Comment (AYAHASSAN287) [reply tests/wrappers_tests/test_send_e2e_part1.py]: done

File: tests/wrappers_tests/test_send_e2e_part1.py (hunk 15 above)
Comment (AYAHASSAN287) [reply tests/wrappers_tests/test_send_e2e_part1.py:63]: added

File: tests/wrappers_tests/test_send_e2e_part1.py (hunk 16 above)
Comment (AYAHASSAN287) [reply tests/wrappers_tests/test_send_e2e_part1.py:630]: comment enhanced

File: tests/wrappers_tests/test_send_e2e_part1.py (hunk 5 above)
Comment (AYAHASSAN287) [reply tests/wrappers_tests/test_send_e2e_part1.py]: now the confirmed failed tests I added xfail and mention the opened issue , for the flaky ones we can leave them to fail without marks , the S20 error u mentioned was addressed in the fixing commits pushed and now we reached a point where the scenario isn't possible to simulate

**Review by at0m1x19 — APPROVED** [review]

**Comment by AYAHASSAN287** [comment]
> LGTM, thanks but can you please add a full run test report to the PR description?

Latest [run](https://github.com/logos-messaging/logos-delivery-interop-tests/actions/runs/25665341173) , all skipped tests failing for issues opened

**Review by fbarbu15 — APPROVED** [review]

File: src/steps/store.py [hunk 17]
```diff
@@ -250,6 +250,62 @@ def check_published_message_is_stored(
...
+                cursor=cursor,
+                page_size=page_size,
+                ascending=ascending,
+                store_v=store_v,
+                **kwargs,
+            )
+
+            logger.debug(f"messages length is {len(self.store_response.messages)}")
+            assert self.store_response.messages, f"Peer {node.image} couldn't find any messages. " f"Actual response: {self.store_response.resp_json}"
+            assert len(self.store_response.messages) >= len(expected_hashes), (
```
Comment (Ivansete-status) [inline src/steps/store.py:301]: Isn't it better to assert strict equal?

File: tests/wrappers_tests/test_send_e2e_part1.py [hunk 18]
```diff
@@ -0,0 +1,1020 @@
...
+        )
+        assert sender_result.is_ok(), f"Failed to start sender: {sender_result.err()}"
+
+        with sender_result.ok_value as sender_node:
+            # relay peer
+            relay_config = {
+                **node_config,
+                "tcpPort": free_port(),
+                "discv5UdpPort": free_port(),
+                "restPort": free_port(),
```
Comment (Ivansete-status) [inline tests/wrappers_tests/test_send_e2e_part1.py:160]: After having merged the following, it should be enough to pass `0` to these to perform auto-port internally. Maybe another test to consider in the future: https://github.com/logos-messaging/logos-delivery/pull/3828

File: tests/wrappers_tests/test_send_e2e_part1.py [hunk 19]
```diff
@@ -0,0 +1,1020 @@
...
+        S21: delivery retry window expires before any valid path recovers.
+        """
+        sender_collector = EventCollector()
+
+        node_config.update(
+            {
+                "relay": True,
+                "store": False,
+                "lightpush": False,
+                "filter": False,
```
Comment (Ivansete-status) [inline tests/wrappers_tests/test_send_e2e_part1.py:364]: Is interesting to only configure nodes by only touching the `mode` attribute, i.e., only setting either `mode: "Edge"` or `mode: "Core"`. Even though nowadays is supported to manually set protocols, we aim for devs to not tweak that directly, in the future.
I mean, this is actually a flaw in the API, that should not accept those attributes, and we might fix that in the future.

**Review by Ivansete-status — COMMENTED** [review]
Sorry for late review. Just adding a few comments. Thanks!

File: tests/wrappers_tests/test_send_e2e_part1.py (hunk 18 above)
Comment (AYAHASSAN287) [reply tests/wrappers_tests/test_send_e2e_part1.py:160]: I will add it in recent tests , add specific test for it

File: tests/wrappers_tests/test_send_e2e_part1.py (hunk 19 above)
Comment (AYAHASSAN287) [reply tests/wrappers_tests/test_send_e2e_part1.py:364]: I considered this in the recent [PR](https://github.com/logos-messaging/logos-delivery-interop-tests/pull/184)

File: src/steps/store.py (hunk 17 above)
Comment (AYAHASSAN287) [reply src/steps/store.py:301]: The real correctness check is the loop (assert expected_hash in actual_hashes)
   Store state isn't isolated ,the store node can hold messages from prior store requests .
